File: typescript_test_crate/tests/01_minimal.ts
class Minimal {
    private _state: string;
    private _state_stack: Array<any>;
    private _state_context: Record<string, any>;
    private _return_value: any;

    constructor() {
        this._state_stack = [];
        this._state_context = {  };
        this._return_value = null;
        this._state = "Start";
        this._enter();
    }

    private _transition(target_state: string, exit_args: any = null, enter_args: any = null) {
        if (exit_args) {
            this._exit(...exit_args);
        } else {
            this._exit();
        }
        this._state = target_state;
        if (enter_args) {
            this._enter(...enter_args);
        } else {
            this._enter();
        }
    }

    private _change_state(target_state: string) {
        this._state = target_state;
    }

    private _dispatch_event(event: string, ...args: any[]) {
        const handler_name = `_s_${this._state}_${event}`;
        const handler = (this as any)[handler_name];
        if (handler) {
            return handler.apply(this, args);
        }
    }

    private _enter(...args: any[]) {
        // No enter handlers
    }

    private _exit(...args: any[]) {
        // No exit handlers
    }

    public is_alive(): boolean {
        this._return_value = null
        this._dispatch_event("is_alive")
        return this._return_value
    }

    private _s_Start_is_alive() {
        this._return_value = true;
        return this._return_value;;
    }
}


function main() {
    console.log("=== Test 01: Minimal System ===");
    const s = new Minimal();

    // Test that system instantiates and responds
    const result = s.is_alive();
    if (result !== true) {
        throw new Error(`Expected true, got ${result}`);
    }
    console.log(`is_alive() = ${result}`);

    console.log("PASS: Minimal system works correctly");
}

main();

File: python_test_crate/tests/01_minimal.py
from typing import Any, Optional, List, Dict, Callable

class Minimal:
    def __init__(self):
        self._state_stack = []
        self._state_context = {}
        self._return_value = None
        self._state = "Start"
        self._enter()

    def _transition(self, target_state, exit_args = None, enter_args = None):
        if exit_args:
            self._exit(*exit_args)
        else:
            self._exit()
        self._state = target_state
        if enter_args:
            self._enter(*enter_args)
        else:
            self._enter()

    def _change_state(self, target_state):
        self._state = target_state

    def _dispatch_event(self, event, *args):
        handler_name = f"_s_{self._state}_{event}"
        handler = getattr(self, handler_name, None)
        if handler:
            return handler(*args)

    def _enter(self, *args):
        # No enter handlers
        pass

    def _exit(self, *args):
        # No exit handlers
        pass

    def is_alive(self) -> bool:
        self._return_value = None
        self._dispatch_event("is_alive")
        return self._return_value

    def _s_Start_is_alive(self) -> bool:
        self._return_value = True
        return


def main():
    print("=== Test 01: Minimal System ===")
    s = Minimal()

    # Test that system instantiates and responds
    result = s.is_alive()
    assert result == True, f"Expected True, got {result}"
    print(f"is_alive() = {result}")

    print("PASS: Minimal system works correctly")

if __name__ == '__main__':
    main()

